README.md
siege
=====

This is a quickly hacked version of siege-2.70 from http://www.joedog.org/siege-home/

A better patch by jsteemann is at https://gist.github.com/2919579



ChangeLog
To email a contributor remove "DELETE" from the email address.
(The DELETEs are necessary as this list is published online.)

2012/01/02 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE RELEASE *        RELEASE_2-72_FEB-15-2012

2012/02/10 Josue Abarca <jmaslibre@NOdebianSPAM.org.gt>
 * SIEGE BETA RELEASE     RELEASE_2-72b1_Feb-10-2012
 * src/version.c          Routine increment 2.72b1
 * src/init.c             Assinged configure generated constant to 
                          the urls.txt file so siege can actually 
                          find it. 

2012/01/02 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE RELEASE *        RELEASE_2-71_FEB-01-2012

2012/30/01 Josue Abarca <jmaslibre@NOdebianSPAM.org.gt>
 * SIEGE BETA RELEASE     RELEASE_2-71b7_Jan-30-2012
 * aclocal.m4             Addressed warnings raised by aclocal-1.9
 * doc/siegerc.in         Changed verbose default to false

2011/03/10 Jason Yan <tailofthesun@NOSPAMgmail.com>
 * SIEGE BETA RELEASE     RELEASE_2-71b6_MAR-10-2011
 * src/http.c             Dynamic Host headers in _get and _post
 * src/util.c             strnlen and strncasestr for folks who lack them
 * src/util.h             prototypes for strnlen and strncasestr

2011/01/25 "Brian Borowski" <brian_borowski@NOSPAMyahoo.com>
 * SIEGE BETA RELEASE     RELEASE_2-71b5_JAN-25-2011
 * lib/notify.c           synchronized memory allotment in message and display

2011/01/24 "Brian Borowski" <brian_borowski@NOSPAMyahoo.com>
 * SIEGE BETA RELEASE     RELEASE_2-71b4_JAN-23-2011
 * lib/notify.c           increased BUFSIZE to match eval.h

2010/12/01 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-71b3_DEC-01-2010

2010/11/30 Christof Damian <christof@NOSPAMdamian.net>
 * src/main.c             terminated the long options
 * src/version.c          increment to 2.71b3

2010/11/22 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-71b2_NOV-22-2010

2010/11/20 Jamie Couture <jamie.couture@NOSPAMgmail.com>
 * src/main.c             Fixed a problem in which the T in Transaction is chomped
 * src/date.c             Added char * timestamp() function
 * src/client.c           Added class to timestamp for stamped output
 * src/setup.h            Added boolean timestamp to my structure
 * src/init.c             Added timestamp=yes|no to .siegerc parser

2010/08/20 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-71b1_AUG-20-2010

2010/08/20 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * src/cookie.c           On line 83, changed 1 to 1UL - bug was 
                          discovered by Toby Matejovsky on MacOS
 * src/version.c          Increment to 2.71b1

2010/07/16 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE RELEASE          RELEASE_2-70_JUL-16-2010

2010/05/21 Josue Abarca <jmaslibre@DELETEdebian.org.gt>
 * utils/bombardment      Fixed a syntax error which caused it to break

2010/05/16 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-70b8_MAY-16-2010
 * src/array.c            dynamic memory allocation
 * src/array.h            updated prototypes; include ARRAYSIZE
 * src/client.c           updated debugging statements in new format
 * src/http.c             updated debugging statements in new format

2010/05/11 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-70b6_MAY-11-2010
 * src/version.c          Increment: 2.70b6
 * src/setup.h            Increased extra to 2048
 * src/http.h             Added the difference (1536) to REQBUF and POSTBUF

2010/05/08 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-70b6_MAY-08-2010
 * lib/notify.c           Added color codings for each level
 * lib/notify.h           Added DEBUG option 
 * include/joedog.h       ADDED DEBUG option
 * src/array.c            Improved memory management
 * src/array.h            Removed size parameter

2010/05/08 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-70b5_MAY-08-2010
 * src/version.c          Increment: 2.70b5
 * src/array.c            Added to distribution (circular array support)
 * src/array.h            Added to distribution (circular array support)
 * src/setup.h            Added ARRAY lurl - array of upto 50 login URLS
 * src/init.c             Added parse login-url and push onto array
 * src/client.c           Loop through login-urls 

2010/04/30 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE     RELEASE_2-70b4_APR-30-2010
 * include/joedog.h       wrapped *nprintf functions with ndef __CYGWIN__
 * src/version.c          Increment: 2.70b4

2010/04/05 Jeffrey Fulmer <http://www.joedog.org/JoeDog/Contact>
 * SIEGE BETA RELEASE *   RELEASE_2-70b3_APR-28-2010
 * src/load.c             Changed detection of binary file types based
                          on file name extensions.

src/url.h
/**
 * URL support
 *
 * Copyright (C) 2000-2006 Jeffrey Fulmer <jeff@joedog.org>
 * This file is part of Siege
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * --
 */
#ifndef  URL_H
#define  URL_H
#include <stdlib.h>
#include <joedog/boolean.h>

#define MAX_URL       4096

/**
 * client data
 */
typedef enum {
  URL_GET   = 0,
  URL_POST  = 1,
  URL_DELETE = 2,
} DIRECTIVE;

/**
 * enum PROTOCOL
 * tcp/ip protocol
 */
typedef enum PROTOCOL{
  HTTP        = 0,
  HTTPS       = 1,
  UNSUPPORTED = 2
} PROTOCOL;


/**
 * URL struct URL
 */
typedef struct
{
  int       urlid;         /* ADDED BY jason, UNIQUE ID   */
  PROTOCOL  protocol;      /* currently only http/https   */
  char      *hostname;     /* DNS entry or IP address     */
  int       port;          /* tcp port num, defs: 80, 443 */
  char      *pathname;     /* path to http resource.      */
  DIRECTIVE calltype;      /* request: GET/POST/HEAD etc. */
  size_t    postlen;       /* length of POST data         */
  char      *postdata;
  char      *posttemp;
  char      *conttype;
  char      url[MAX_URL];
  time_t    expires;
  time_t    modified;
  BOOLEAN   cached;
  char      *etag;
} URL;

int      protocol_length(char *url); 
BOOLEAN  is_supported(char* url);
int      get_default_port(PROTOCOL p);
PROTOCOL get_protocol(const char *url);
void     insert_childid(URL *U, int mypid);
void     url_set_last_modified(URL *U, char *date);
void     url_set_etag(URL *U, char *etag);
void     url_set_expires(URL *U, int secs);
char     *url_get_if_modified_since(URL *U);
char     *url_get_etag(URL *U);
URL      *build_url(char *url, int defaultport, int id);
URL      *add_url(char *url, int id);

#endif/*URL_H*/


src/http.h
/**
 * SIEGE http header file
 *
 * Copyright (C) 2000, 2001, 2002 by
 * Jeffrey Fulmer - <jdfulmer@armstrong.com>
 * This file is distributed as part of Siege 
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
#ifndef HTTP_H
#define HTTP_H

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <sock.h>
#include <cookie.h>
#include <auth.h>
#include <url.h>
#include <joedog/joedog.h>
#include <joedog/boolean.h>

#ifndef HAVE_SNPRINTF
# define portable_snprintf  snprintf
# define portable_vsnprintf vsnprintf
#endif

#define REQBUF  43008
#define POSTBUF 63488

/**
 * HTTP headers structure.
 */
typedef 
struct headers
{
  char                head[64];                /* http response         */
  int                 code;                    /* http return code      */
  unsigned long int   length;
  char                cookie[MAX_COOKIE_SIZE]; /* set-cookie data       */
  char                *redirect;               /* redirection URL       */
  struct{
    int   www;                                 /* www auth challenge    */
    int   proxy;                               /* proxy auth challenge  */
    struct {
      char  *www;                              /* www auth realm        */
      char  *proxy;                            /* proxy auth realm      */
    } realm;
    struct {
      char  *www;                              /* www auth realm        */
      char  *proxy;                            /* proxy auth realm      */
    } challenge;                               /* square peg/round hole */
    struct {
      TYPE  www;
      TYPE  proxy;
    } type;
  } auth;
  int                 keepalive;               /* http keep-alive       */
} HEADERS;

/* http function prototypes */
int      http_get (CONN *C, URL *U);
int      http_post(CONN *C, URL *U);
int      http_delete (CONN *C, URL *U);

void     http_free_headers(HEADERS *h);
HEADERS *http_read_headers(CONN *C, URL *U);
ssize_t  http_read(CONN *C);
BOOLEAN  https_tunnel_request(CONN *C, char *host, int port);
int      https_tunnel_response(CONN *C);

#endif /* HTTP_H */



src/init.c
/**
 * Siege environment initialization.
 *
 * Copyright (C) 2000-2010 by
 * Jeffrey Fulmer - <jeff@joedog.org>, et al. 
 * This file is distributed as part of Siege 
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */  
#include <init.h>
#include <setup.h>
#include <auth.h>
#include <util.h>
#include <hash.h>
#include <eval.h>
#include <fcntl.h>
#include <version.h>
#include <joedog/boolean.h>
#include <joedog/defs.h>
#include <joedog/joedog.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>

int
init_config( void )
{
  char *e;
  int  res;
  struct stat buf;

  /* Check if we were passed the -R switch to use a different siegerc file.
   * If not, check for the presence of the SIEGERC variable, otherwise
   * use default of ~/.siegerc */
  if(strcmp(my.rc, "") == 0){
    if((e = getenv("SIEGERC")) != NULL){
      snprintf(my.rc, sizeof(my.rc), e);
    } else {
      snprintf(my.rc, sizeof(my.rc), "%s/.siegerc", getenv("HOME"));
      if (stat(my.rc, &buf) < 0 && errno == ENOENT) {
        snprintf(my.rc, sizeof(my.rc), CNF_FILE);
      }
    }
  } 

  my.debug          = FALSE;
  my.internet       = FALSE;
  my.config         = FALSE;
  my.cookies        = TRUE;
  my.csv            = FALSE;
  my.fullurl        = FALSE;
  my.secs           = -1;
  my.reps           = MAXREPS; 
  my.bids           = 5;
  my.login          = FALSE;
  my.failures       = 1024;
  my.failed         = 0;
  my.auth.encode    = xstrdup("");
  my.proxy.encode   = xstrdup("");
  my.proxy.required = FALSE;
  my.proxy.port     = 80; 
  my.timeout        = 30;
  my.timestamp      = FALSE;
  my.chunked        = FALSE;
  my.extra[0]       = 0;
  my.follow         = TRUE;
  my.zero_ok        = TRUE; 
  my.signaled       = 0;
  my.ssl_timeout    = 300;
  my.ssl_cert       = NULL;
  my.ssl_key        = NULL;
  my.ssl_ciphers    = NULL; 
  my.lurl           = new_array();

  if((res = pthread_mutex_init(&(my.lock), NULL)) != 0)
    NOTIFY(FATAL, "unable to initiate lock");
  if((res = pthread_cond_init(&(my.cond), NULL )) != 0)
    NOTIFY(FATAL, "unable to initiate condition");

  if(load_conf(my.rc) < 0){
    fprintf( stderr, "****************************************************\n" );
    fprintf( stderr, "siege: could not open %s\n", my.rc );
    fprintf( stderr, "run \'siege.config\' to generate a new .siegerc file\n" );
    fprintf( stderr, "****************************************************\n" );
    return -1;
  }
  
  if(strlen(my.file) < 1){ 
    snprintf(
      my.file, sizeof( my.file ),
      "%s", URL_FILE
    );
  }

  if(strlen(my.uagent) < 1) 
    snprintf( 
      my.uagent, sizeof(my.uagent),
      "JoeDog/1.00 [en] (X11; I; Siege %s)", version_string 
    );

  if(strlen(my.encoding) < 1)
    snprintf(
      my.encoding, sizeof(my.encoding), "*"
    );

  if(strlen(my.logfile) < 1) 
    snprintf( 
      my.logfile, sizeof( my.logfile ),
      "%s", LOG_FILE 
    );
  /**
   * DEPRECATED!! username and password are being
   * phased out in favor of my.auth.head
   */
  if(( my.username && strlen(my.username) > 0 ) &&
    (  my.password && strlen(my.password) > 0 )){
    add_authorization(WWW, my.username, my.password, "all"); 
  }
  if(my.proxy.hostname && strlen(my.proxy.hostname) > 0){
    my.proxy.required = TRUE;
  } 

  return 0;  
}

int
show_config( int EXIT )
{
  printf( "CURRENT  SIEGE  CONFIGURATION\n" );
  printf( "%s\n", my.uagent ); 
  printf( "Edit the resource file to change the settings.\n" );
  printf( "----------------------------------------------\n" );
  printf( "version:                        %s\n", version_string );
  printf( "verbose:                        %s\n", my.verbose?"true":"false" );
  printf( "debug:                          %s\n", my.debug?"true":"false" );
  printf( "protocol:                       %s\n", my.protocol?"HTTP/1.1":"HTTP/1.0" );
  if(my.proxy.required){
    printf("proxy-host:                     %s\n", my.proxy.hostname);
    printf("proxy-port:                     %d\n", my.proxy.port);
  }
  printf( "connection:                     %s\n", my.keepalive?"keep-alive":"close" );
  printf( "concurrent users:               %d\n", my.cusers );
  if( my.secs > 0 )
    printf( "time to run:                    %d seconds\n", my.secs );
  else
    printf( "time to run:                    n/a\n" );
  if(( my.reps > 0 )&&( my.reps != MAXREPS ))
    printf( "repetitions:                    %d\n", my.reps );
  else
    printf( "repetitions:                    n/a\n" );
  printf( "socket timeout:                 %d\n", my.timeout );
  printf( "delay:                          %d sec%s\n", my.delay,my.delay>1?"s":"" );
  printf( "internet simulation:            %s\n", my.internet?"true":"false"  );
  printf( "benchmark mode:                 %s\n", my.bench?"true":"false"  );
  printf( "failures until abort:           %d\n", my.failures );
  printf( "named URL:                      %s\n", my.url==NULL||strlen(my.url)<2?"none":my.url );
  printf( "URLs file:                      %s\n", strlen(my.file)>1?my.file:URL_FILE );
  printf( "logging:                        %s\n", my.logging?"true":"false" );
  printf( "log file:                       %s\n", my.logfile==NULL?LOG_FILE:my.logfile );
  printf( "resource file:                  %s\n", my.rc);
  printf( "timestamped output:             %s\n", my.timestamp?"true":"false");
  printf( "comma separated output:         %s\n", my.csv?"true":"false");
  printf( "allow redirects:                %s\n", my.follow?"true":"false" );
  printf( "allow zero byte data:           %s\n", my.zero_ok?"true":"false" ); 
  printf( "allow chunked encoding:         %s\n", my.chunked?"true":"false" ); 
  printf( "proxy auth:                     " ); display_authorization( PROXY );printf( "\n" );
  printf( "www auth:                       " ); display_authorization( WWW ); 
  printf( "\n" );

  if (EXIT) exit(0);
  else return 0;
}

static char
*get_line(FILE *fp)
{
  char *ptr = NULL;
  char *newline;
  char tmp[256];

  memset( tmp, 0, sizeof(tmp)); 
  do {
    if((fgets(tmp, sizeof(tmp), fp)) == NULL) return(NULL);
    if(ptr == NULL) {
      ptr = xstrdup( tmp );
    } else {
      ptr = (char*)xrealloc(ptr, strlen(ptr) + strlen(tmp) + 1);
      strcat( ptr, tmp );
    }
    newline = strchr(ptr, '\n');
  } while( newline == NULL );
  *newline = '\0';
 
  return ptr;
} 

static char
*chomp_line(FILE *fp, char **mystr, int *line_num)
{
  char *ptr;
  while(TRUE){
    if((*mystr = get_line( fp )) == NULL) return NULL;
    (*line_num)++;
    ptr = chomp(*mystr);
    /* exclude comments */
    if(*ptr != '#' && *ptr != '\0'){
      return(ptr);
    } else {
      xfree(ptr);
    }
  }
} 

int
load_conf(char *filename)
{
  FILE *fp;
  HASH H;
  int  line_num = 0;
  char *line;
  char *tmp;
  char *option;
  char *optionptr;
  char *value;
 
  if ((fp = fopen(filename, "r")) == NULL) {
    return -1;
  } 

  H = new_hash(16);

  while((line = chomp_line(fp, &line, &line_num)) != NULL){
    tmp    = line;
    optionptr = option = xstrdup(line);
    while(*optionptr && !ISSPACE((int)*optionptr) && !ISSEPARATOR(*optionptr))
      optionptr++;
    *optionptr++=0;
    while(ISSPACE((int)*optionptr) || ISSEPARATOR(*optionptr))
      optionptr++;
    value  = xstrdup(optionptr);
    while(*line)
      line++;  
    while(strstr(option, "$")){
      option = evaluate(H, option);
    }
    while(strstr(value, "$")){
      value = evaluate(H, value);
    } 
    if(strmatch(option, "verbose")){
      if(!strncasecmp(value, "true", 4))
        my.verbose = TRUE;
      else
        my.verbose = FALSE;
    } 
    else if(strmatch(option, "csv")){
      if(!strncasecmp(value, "true", 4))
        my.csv = TRUE;
      else
        my.csv = FALSE;
    } 
    else if(strmatch(option, "fullurl")){
      if(!strncasecmp(value, "true", 4))
        my.fullurl = TRUE;
      else
        my.fullurl = FALSE;
    } 
    else if(strmatch(option, "display-id")){
      if(!strncasecmp(value, "true", 4))
        my.display = TRUE;
      else
        my.display = FALSE;
    } 
    else if(strmatch( option, "logging")){
      if(!strncasecmp( value, "true", 4))
        my.logging = TRUE;
      else
        my.logging = FALSE;
    }
    else if(strmatch(option, "show-logfile")){
      if(!strncasecmp(value, "true", 4))
        my.shlog = TRUE;
      else
        my.shlog = FALSE;
    }
    else if(strmatch(option, "logfile")){
      strncpy(my.logfile, value, sizeof(my.logfile)); 
    } 
    else if(strmatch(option, "cookies")){
      if(strmatch(value, "true"))
        my.cookies = TRUE;
      else
        my.cookies = FALSE;
    }
    else if(strmatch(option, "concurrent")){
      if(value != NULL){
        my.cusers = atoi(value);
      } else {
        my.cusers = 10;
      }
    } 
    else if(strmatch(option, "reps")){
      if(value != NULL){
        my.reps = atoi(value);
      } else {
        my.reps = 5;
      }
    }
    else if(strmatch(option, "time")){
      parse_time(value);
    }
    else if(strmatch(option, "delay")){
      if(value != NULL){
        my.delay = atoi(value);
      } else {
        my.delay = 1;
      }
    }
    else if(strmatch(option, "timeout")){
      if(value != NULL){
        my.timeout = atoi(value);
      } else {
        my.timeout = 15;
      }
    }
    else if(strmatch(option, "timestamp")){
      if(!strncasecmp(value, "true", 4)) 
        my.timestamp = TRUE;
      else
        my.timestamp = FALSE;
    }
    else if(strmatch(option, "internet")){
      if(!strncasecmp(value, "true", 4))
        my.internet = TRUE;
      else
        my.internet = FALSE;
    }
    else if(strmatch(option, "benchmark")){
      if(!strncasecmp(value, "true", 4)) 
        my.bench = TRUE;
      else
        my.bench = FALSE;
    }
    else if(strmatch(option, "cache")){
      if(!strncasecmp(value, "true", 4)) 
        my.cache = TRUE;
      else
        my.cache = FALSE;
    }
    else if(strmatch( option, "debug")){
      if(!strncasecmp( value, "true", 4))
        my.debug = TRUE;
      else
        my.debug = FALSE;
    }
    else if(strmatch(option, "file")){
      memset(my.file, 0, sizeof(my.file));
      strncpy(my.file, value, sizeof(my.file));
    }
    else if(strmatch(option, "url")){
      my.url = stralloc(value);
    }
    else if(strmatch(option, "user-agent")){
      strncpy(my.uagent, value, sizeof(my.uagent));
    }
    else if(strmatch(option, "accept-encoding")){
      strncpy(my.encoding, value, sizeof(my.encoding));
    }
    else if(!strncasecmp(option, "login", 5)){
      if(strmatch(option, "login-url")){  
        /* login URL */
        // XXX: my.loginurl = stralloc(value); 
        my.login = TRUE;
        array_push(my.lurl, value);
      } else {
        /* user login info */
        char *usr, *pwd, *rlm, *tmpvalue;
        usr = tmpvalue = value;
        while( *tmpvalue && *tmpvalue != ':' && *tmpvalue != '\0' )
          tmpvalue++;
        *tmpvalue++=0; pwd = tmpvalue;
        while( *tmpvalue && *tmpvalue != ':' && *tmpvalue != '\0' )
	  tmpvalue++;
	if('\0' != *tmpvalue) {
	  *tmpvalue++=0;
	  rlm = tmpvalue;
	} else {
	  rlm = NULL;
	}
        add_authorization(WWW, usr, pwd, rlm);
      }
    }
    else if(strmatch(option, "attempts")){
      if(value != NULL){
        my.bids = atoi(value);
      } else { 
        my.bids = 3;
      }
    }
    else if(strmatch(option, "username")){
      my.username = stralloc(trim(value));
    }
    else if(strmatch(option, "password")){
      my.password = stralloc(trim(value));
    }
    else if(strmatch(option, "connection")){
      if(!strncasecmp(value, "keep-alive", 10))
        my.keepalive = TRUE;
      else
        my.keepalive = FALSE; 
    }
    else if(strmatch(option, "protocol")){
      if(!strncasecmp(value, "HTTP/1.1", 8))
        my.protocol = TRUE;
      else
        my.protocol = FALSE; 
    }
    else if(strmatch(option, "proxy-host")){
      my.proxy.hostname = xstrdup(trim(value));
    }
    else if(strmatch(option, "proxy-port")){
      if(value != NULL){
        my.proxy.port = atoi(value);
      } else {
        my.proxy.port = 3128;
      }
    } 
    else if(strmatch(option, "proxy-login")){
      char *usr, *pwd, *rlm, *tmpvalue;
      usr = tmpvalue = value;
      while(*tmpvalue && *tmpvalue != ':' && *tmpvalue != '\0')
        tmpvalue++;
      *tmpvalue++=0; pwd = tmpvalue;
      while(*tmpvalue && *tmpvalue != ':' && *tmpvalue != '\0')
      tmpvalue++;
      if('\0' != *tmpvalue) {
	*tmpvalue++=0;
	rlm = tmpvalue;
      } else {
        rlm = NULL;
      }
      add_authorization(PROXY, usr, pwd, rlm);  
    }
    else if(strmatch(option, "failures")){
      if(value != NULL){
        my.failures = atoi(value);
      } else {
        my.failures = 30;
      }
    }
    else if(strmatch(option, "chunked")){
      if(!strncasecmp(value, "true", 4))
        my.chunked = TRUE;
      else
        my.chunked = FALSE;  
    }
    else if(strmatch(option, "header")){
      if(!strchr(value,':')) NOTIFY(FATAL, "no ':' in http-header");
      if((strlen(value) + strlen(my.extra) + 3) > 512) NOTIFY(FATAL, "too many headers");
      strcat(my.extra,value);
      strcat(my.extra,"\015\012");
    }
    else if(strmatch(option, "expire-session")){
      if (!strncasecmp(value, "true", 4 ))
        my.expire = TRUE;
      else
        my.expire = FALSE;
    }
    else if(strmatch(option, "follow-location")){
      if ( !strncasecmp( value, "true", 4 ))
        my.follow = TRUE;
      else
        my.follow = FALSE;
    }
    else if(strmatch(option, "zero-data-ok")){
      if( !strncasecmp(value, "true", 4))
        my.zero_ok = TRUE;
      else
        my.zero_ok = FALSE;
    } 
    else if(strmatch(option, "ssl-cert")){
      my.ssl_cert = stralloc(value);
    }
    else if(strmatch(option, "ssl-key")){
      my.ssl_key = stralloc(value);
    }
    else if(strmatch(option, "ssl-timeout")){
      if(value != NULL){
        my.ssl_timeout = atoi(value);
      } else {
        my.ssl_timeout = 15;
      }
    }
    else if(strmatch(option, "ssl-ciphers")){
      my.ssl_ciphers = stralloc(value);
    } 
    else if(strmatch(option, "spinner")){
      if(!strncasecmp(value, "true", 4))
        my.spinner = TRUE;
      else
        my.spinner = FALSE;
    } else {
      hash_add(H, option, value);
    }
    xfree(value);
    xfree(option);
    free(tmp);
  } /* end of while line=chomp_line */

  hash_destroy(H);
  fclose(fp);
  return 0;
}

/**
 * don't be insulted, the author is the 
 * DS Module in question...   ;-)
 */ 
void
ds_module_check( void )
{
  if( my.bench ){ 
#if defined(hpux) || defined(__hpux)
    my.delay = 1; 
#else
    my.delay = 0; 
#endif
  }
  if( my.secs > 0 && (( my.reps > 0 ) && ( my.reps != MAXREPS ))){
    NOTIFY(ERROR, "CONFIG conflict: selected time and repetition based testing" );
    fprintf( stderr, "defaulting to time-based testing: %d seconds\n", my.secs );
    my.reps = MAXREPS;
  }
}

